updatePolicies is not supported in redis-adapter

Hi, When I am trying to use `updatePolicies` API it gives me the following exception:

```
panic: interface conversion: *redisadapter.Adapter is not persist.UpdatableAdapter: missing method UpdateFilteredPolicies

goroutine 1 [running]:
github.com/casbin/casbin/v2.(*Enforcer).updatePolicies(0xc000144400, {0x6bef5b, 0x1}, {0x6bef5b, 0x1}, {0xc00015c540, 0x2, 0x2}, {0xc00015c570, 0x2, ...})
	/root/go/pkg/mod/github.com/casbin/casbin/v2@v2.51.0/internal_api.go:204 +0x11a
github.com/casbin/casbin/v2.(*Enforcer).UpdateNamedPolicies(...)
	/root/go/pkg/mod/github.com/casbin/casbin/v2@v2.51.0/management_api.go:253
github.com/casbin/casbin/v2.(*Enforcer).UpdatePolicies(0xc00013dd58?, {0xc00015c540?, 0x6bef5b?, 0x1?}, {0xc00015c570?, 0xc000024240?, 0x20?})
	/root/go/pkg/mod/github.com/casbin/casbin/v2@v2.51.0/management_api.go:249 +0x4f
main.(*Role).updatepolicy(0xc00013de18, {0x6bff99, 0x8}, {0x6bf711, 0x6}, 0x0)
	

```

--- a/adapter.go
+++ b/adapter.go
@@ -15,9 +15,13 @@
 package redisadapter
 
 import (
+	"bytes"
 	"encoding/json"
 	"errors"
+	"fmt"
+	"regexp"
 	"runtime"
+	"strings"
 
 	"github.com/casbin/casbin/v2/model"
 	"github.com/casbin/casbin/v2/persist"
@@ -26,22 +30,23 @@ import (
 
 // CasbinRule is used to determine which policy line to load.
 type CasbinRule struct {
-	PType string `xorm:"varchar(100) index"`
-	V0    string `xorm:"varchar(100) index"`
-	V1    string `xorm:"varchar(100) index"`
-	V2    string `xorm:"varchar(100) index"`
-	V3    string `xorm:"varchar(100) index"`
-	V4    string `xorm:"varchar(100) index"`
-	V5    string `xorm:"varchar(100) index"`
+	PType string
+	V0    string
+	V1    string
+	V2    string
+	V3    string
+	V4    string
+	V5    string
 }
 
 // Adapter represents the Redis adapter for policy storage.
 type Adapter struct {
-	network  string
-	address  string
-	key      string
-	password string
-	conn     redis.Conn
+	network    string
+	address    string
+	key        string
+	password   string
+	conn       redis.Conn
+	isFiltered bool
 }
 
 // finalizer is the destructor for Adapter.
@@ -146,28 +151,37 @@ func (a *Adapter) createTable() {
 }
 
 func (a *Adapter) dropTable() {
+	_, _ = a.conn.Do("DEL", a.key)
+}
+
+func (c *CasbinRule) toStringPolicy() []string {
+	policy := make([]string, 0)
+	if c.PType != "" {
+		policy = append(policy, c.PType)
+	}
+	if c.V0 != "" {
+		policy = append(policy, c.V0)
+	}
+	if c.V1 != "" {
+		policy = append(policy, c.V1)
+	}
+	if c.V2 != "" {
+		policy = append(policy, c.V2)
+	}
+	if c.V3 != "" {
+		policy = append(policy, c.V3)
+	}
+	if c.V4 != "" {
+		policy = append(policy, c.V4)
+	}
+	if c.V5 != "" {
+		policy = append(policy, c.V5)
+	}
+	return policy
 }
 
 func loadPolicyLine(line CasbinRule, model model.Model) {
-	text := []string{line.PType}
-	if line.V0 != "" {
-		text = append(text, line.V0)
-	}
-	if line.V1 != "" {
-		text = append(text, line.V1)
-	}
-	if line.V2 != "" {
-		text = append(text, line.V2)
-	}
-	if line.V3 != "" {
-		text = append(text, line.V3)
-	}
-	if line.V4 != "" {
-		text = append(text, line.V4)
-	}
-	if line.V5 != "" {
-		text = append(text, line.V5)
-	}
+	text := line.toStringPolicy()
 
 	persist.LoadPolicyArray(text, model)
 }
@@ -199,6 +213,7 @@ func (a *Adapter) LoadPolicy(model model.Model) error {
 		loadPolicyLine(line, model)
 	}
 
+	a.isFiltered = false
 	return nil
 }
 
@@ -283,11 +298,6 @@ func (a *Adapter) RemovePolicy(sec string, ptype string, rule []string) error {
 	return err
 }
 
-// RemoveFilteredPolicy removes policy rules that match the filter from the storage.
-func (a *Adapter) RemoveFilteredPolicy(sec string, ptype string, fieldIndex int, fieldValues ...string) error {
-	return errors.New("not implemented")
-}
-
 // AddPolicies adds policy rules to the storage.
 func (a *Adapter) AddPolicies(sec string, ptype string, rules [][]string) error {
 	var texts [][]byte
@@ -317,4 +327,304 @@ func (a *Adapter) RemovePolicies(sec string, ptype string, rules [][]string) err
 		}
 	}
 	return nil
+}
+
+//FilteredAdapter
+
+// IsFiltered returns true if the loaded policy has been filtered.
+func (a *Adapter) IsFiltered() bool {
+	return a.isFiltered
+}
+
+type Filter struct {
+	PType []string
+	V0    []string
+	V1    []string
+	V2    []string
+	V3    []string
+	V4    []string
+	V5    []string
+}
+
+func filterToRegexPattern(filter *Filter) string {
+	// example data in redis: {"PType":"p","V0":"data2_admin","V1":"data2","V2":"write","V3":"","V4":"","V5":""}
+
+	var f = [][]string{filter.PType,
+		filter.V0, filter.V1, filter.V2,
+		filter.V3, filter.V4, filter.V5}
+
+	args := []interface{}{}
+	for _, v := range f {
+		if len(v) == 0 {
+			args = append(args, ".*")
+		} else {
+			escapedV := make([]string, 0, len(v))
+			for _, s := range v {
+				escapedV = append(escapedV, regexp.QuoteMeta(s))
+			}
+			args = append(args, "(?:"+strings.Join(escapedV, "|")+")") // (?:data2_admin|data1_admin)
+		}
+	}
+
+	// example pattern:
+	//^\{"PType":".*","V0":"(?:data2_admin|data1_admin)","V1":".*","V2":".*","V3":".*","V4":".*","V5":".*"\}$
+	pattern := fmt.Sprintf(
+		`^\{"PType":"%s","V0":"%s","V1":"%s","V2":"%s","V3":"%s","V4":"%s","V5":"%s"\}$`, args...,
+	)
+	return pattern
+}
+
+func escapeLuaPattern(s string) string {
+	var buf bytes.Buffer
+	for _, char := range s {
+		switch char {
+		case '.', '%', '-', '+', '*', '?', '^', '$', '(', ')', '[', ']': // magic chars: . % + - * ? [ ( ) ^ $
+			buf.WriteRune('%')
+		}
+		buf.WriteRune(char)
+	}
+	return buf.String()
+}
+
+func filterFieldToLuaPattern(sec string, ptype string, fieldIndex int, fieldValues ...string) string {
+	args := []interface{}{ptype}
+
+	idx := fieldIndex + len(fieldValues)
+	for i := 0; i < 6; i++ { // v0-v5
+		if fieldIndex <= i && idx > i && fieldValues[i-fieldIndex] != "" {
+			args = append(args, escapeLuaPattern(fieldValues[i-fieldIndex]))
+		} else {
+			args = append(args, ".*")
+		}
+	}
+
+	// example pattern:
+	// ^{"PType":"p","V0":"data2_admin","V1":".*","V2":".*","V3":".*","V4":".*","V5":".*"}$
+	pattern := fmt.Sprintf(
+		`^{"PType":"%s","V0":"%s","V1":"%s","V2":"%s","V3":"%s","V4":"%s","V5":"%s"}$`, args...,
+	)
+	return pattern
+}
+
+func (a *Adapter) loadFilteredPolicy(model model.Model, filter *Filter) error {
+	num, err := redis.Int(a.conn.Do("LLEN", a.key))
+	if err == redis.ErrNil {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	values, err := redis.Values(a.conn.Do("LRANGE", a.key, 0, num))
+	if err != nil {
+		return err
+	}
+
+	re := regexp.MustCompile(filterToRegexPattern(filter))
+
+	var line CasbinRule
+	for _, value := range values {
+		text, ok := value.([]byte)
+		if !ok {
+			return errors.New("the type is wrong")
+		}
+
+		if !re.Match(text) {
+			continue
+		}
+
+		err = json.Unmarshal(text, &line)
+		if err != nil {
+			return err
+		}
+		loadPolicyLine(line, model)
+	}
+	return nil
+}
+
+// LoadFilteredPolicy loads only policy rules that match the filter.
+func (a *Adapter) LoadFilteredPolicy(model model.Model, filter interface{}) error {
+	if filter == nil {
+		return a.LoadPolicy(model)
+	}
+
+	var err error
+	switch f := filter.(type) {
+	case *Filter:
+		err = a.loadFilteredPolicy(model, f)
+	case Filter:
+		err = a.loadFilteredPolicy(model, &f)
+	default:
+		err = fmt.Errorf("invalid filter type")
+	}
+
+	if err != nil {
+		return err
+	}
+	a.isFiltered = true
+	return nil
+}
+
+// RemoveFilteredPolicy removes policy rules that match the filter from the storage.
+func (a *Adapter) RemoveFilteredPolicy(sec string, ptype string, fieldIndex int, fieldValues ...string) error {
+
+	pattern := filterFieldToLuaPattern(sec, ptype, fieldIndex, fieldValues...)
+
+	var getScript = redis.NewScript(1, `
+		local key = KEYS[1]
+		local pattern = ARGV[1]
+		
+		local r = redis.call('lrange', key, 0, -1)
+		for i=1, #r do 
+			if  string.find(r[i], pattern) then
+				redis.call('lset', key, i-1, '__CASBIN_DELETED__')
+			end
+		end
+		redis.call('lrem', key, 0, '__CASBIN_DELETED__')
+		return 
+	`)
+	_, err := getScript.Do(a.conn, a.key, pattern)
+	return err
+}
+
+// UpdatableAdapter
+
+// UpdatePolicy updates a new policy rule to DB.
+func (a *Adapter) UpdatePolicy(sec string, ptype string, oldRule, newPolicy []string) error {
+	oldLine := savePolicyLine(ptype, oldRule)
+	textOld, err := json.Marshal(oldLine)
+	if err != nil {
+		return err
+	}
+	newLine := savePolicyLine(ptype, newPolicy)
+	textNew, err := json.Marshal(newLine)
+	if err != nil {
+		return err
+	}
+
+	var getScript = redis.NewScript(1, `
+		local key = KEYS[1]
+		local old = ARGV[1]
+		local newRule = ARGV[2]
+	
+		local r = redis.call('lrange', key, 0, -1)
+		for i=1,#r do
+			if r[i] == old then
+				redis.call('lset', key, i-1, newRule)
+				return true
+			end
+		end
+		return false
+	`)
+	_, err = getScript.Do(a.conn, a.key, textOld, textNew)
+	return err
+}
+
+func (a *Adapter) UpdatePolicies(sec string, ptype string, oldRules, newRules [][]string) error {
+
+	if len(oldRules) != len(newRules) {
+		return errors.New("oldRules and newRules should have the same length")
+	}
+
+	oldPolicies := make([]string, 0, len(oldRules))
+	newPolicies := make([]string, 0, len(newRules))
+	for _, oldRule := range oldRules {
+		textOld, err := json.Marshal(savePolicyLine(ptype, oldRule))
+		if err != nil {
+			return err
+		}
+		oldPolicies = append(oldPolicies, string(textOld))
+	}
+	for _, newRule := range newRules {
+		textNew, err := json.Marshal(savePolicyLine(ptype, newRule))
+		if err != nil {
+			return err
+		}
+		newPolicies = append(newPolicies, string(textNew))
+	}
+
+	// Initialize a package-level variable with a script.
+	var getScript = redis.NewScript(1, `
+		local key = KEYS[1]
+		local len = #ARGV/2
+		
+		local map = {}
+		for i = 1, len, 1 do
+			map[ARGV[i]] = ARGV[i + len] -- map[oldRule] = newRule
+		end
+		
+		local r = redis.call('lrange', key, 0, -1)
+		for i=1,#r do
+			if map[r[i]] ~= nil then
+				redis.call('lset', key, i-1, map[r[i]])
+				-- return true
+			end
+		end
+		
+		return false
+	`)
+	args := redis.Args{}.Add(a.key).AddFlat(oldPolicies).AddFlat(newPolicies)
+	_, err := getScript.Do(a.conn, args...)
+	return err
+}
+
+func (a *Adapter) UpdateFilteredPolicies(sec string, ptype string, newPolicies [][]string, fieldIndex int, fieldValues ...string) ([][]string, error) {
+	// UpdateFilteredPolicies deletes old rules and adds new rules.
+
+	oldP := make([]string, 0)
+	newP := make([]string, 0, len(newPolicies))
+	for _, newRule := range newPolicies {
+		textNew, err := json.Marshal(savePolicyLine(ptype, newRule))
+		if err != nil {
+			return nil, err
+		}
+		newP = append(newP, string(textNew))
+	}
+
+	pattern := filterFieldToLuaPattern(sec, ptype, fieldIndex, fieldValues...)
+
+	// Initialize a package-level variable with a script.
+	var getScript = redis.NewScript(1, `
+		local key = KEYS[1]
+		local pattern = ARGV[1]
+		
+		local ret = {}
+		local r = redis.call('lrange', key, 0, -1)
+		for i=1, #r do 
+			if  string.find(r[i], pattern) then
+        		table.insert(ret, r[i])
+				redis.call('lset', key, i-1, '__CASBIN_DELETED__')
+			end
+		end
+		redis.call('lrem', key, 0, '__CASBIN_DELETED__')
+		
+		local r = redis.call('lrange', key, 0, -1)
+		for i=2,#r do
+			redis.call('rpush', key, ARGV[i])
+		end
+		
+		return ret
+	`)
+	args := redis.Args{}.Add(a.key).Add(pattern).AddFlat(newP)
+	//r, err := getScript.Do(a.conn, args...)
+	//reply, err := redis.Values(r, err)
+	reply, err := redis.Values(getScript.Do(a.conn, args...))
+	if err != nil {
+		return nil, err
+	}
+
+	if err = redis.ScanSlice(reply, &oldP); err != nil {
+		return nil, err
+	}
+
+	ret := make([][]string, 0, len(oldP))
+	for _, oldRule := range oldP {
+		var line CasbinRule
+		if err := json.Unmarshal([]byte(oldRule), &line); err != nil {
+			return nil, err
+		}
+
+		ret = append(ret, line.toStringPolicy())
+	}
+
+	return ret, nil
 }

--- a/adapter_test.go
+++ b/adapter_test.go
@@ -16,6 +16,7 @@ package redisadapter
 
 import (
 	"log"
+	"strings"
 	"testing"
 
 	"github.com/casbin/casbin/v2"
@@ -27,85 +28,340 @@ func testGetPolicy(t *testing.T, e *casbin.Enforcer, res [][]string) {
 	myRes := e.GetPolicy()
 	log.Print("Policy: ", myRes)
 
-	if !util.Array2DEquals(res, myRes) {
-		t.Error("Policy: ", myRes, ", supposed to be ", res)
+	m := make(map[string]bool, len(res))
+	for _, value := range res {
+		key := strings.Join(value, ",")
+		m[key] = true
+	}
+
+	for _, value := range myRes {
+		key := strings.Join(value, ",")
+		if !m[key] {
+			t.Error("Policy: ", myRes, ", supposed to be ", res)
+			break
+		}
 	}
 }
 
-func TestAdapter(t *testing.T) {
+func initPolicy(t *testing.T, a *Adapter) {
 	// Because the DB is empty at first,
 	// so we need to load the policy from the file adapter (.CSV) first.
 	e, _ := casbin.NewEnforcer("examples/rbac_model.conf", "examples/rbac_policy.csv")
 
-	a := NewAdapter("tcp", "127.0.0.1:6379")
-	// Use the following if Redis has password like "123"
-	//a := NewAdapterWithPassword("tcp", "127.0.0.1:6379", "123")
-	t.Run("Read the policies from an empty redis", func(t *testing.T) {
-		if err := a.LoadPolicy(e.GetModel()); err != nil {
-			t.Error("Should not return an error")
-		}
-	})
-
 	// This is a trick to save the current policy to the DB.
 	// We can't call e.SavePolicy() because the adapter in the enforcer is still the file adapter.
 	// The current policy means the policy in the Casbin enforcer (aka in memory).
-	a.SavePolicy(e.GetModel())
+	err := a.SavePolicy(e.GetModel())
+	if err != nil {
+		panic(err)
+	}
 
 	// Clear the current policy.
 	e.ClearPolicy()
 	testGetPolicy(t, e, [][]string{})
 
 	// Load the policy from DB.
-	a.LoadPolicy(e.GetModel())
+	err = a.LoadPolicy(e.GetModel())
+	if err != nil {
+		panic(err)
+	}
 	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}})
+}
 
+func testSaveLoad(t *testing.T, a *Adapter) {
+	// Initialize some policy in DB.
+	initPolicy(t, a)
 	// Note: you don't need to look at the above code
 	// if you already have a working DB with policy inside.
 
 	// Now the DB has policy, so we can provide a normal use case.
 	// Create an adapter and an enforcer.
 	// NewEnforcer() will load the policy automatically.
-	a = NewAdapter("tcp", "127.0.0.1:6379")
+	e, _ := casbin.NewEnforcer("examples/rbac_model.conf", a)
+	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}})
+}
+
+func testAutoSave(t *testing.T, a *Adapter) {
+	// Initialize some policy in DB.
+	initPolicy(t, a)
+	// Note: you don't need to look at the above code
+	// if you already have a working DB with policy inside.
+
+	// Now the DB has policy, so we can provide a normal use case.
+	// Create an adapter and an enforcer.
+	// NewEnforcer() will load the policy automatically.
+	e, _ := casbin.NewEnforcer("examples/rbac_model.conf", a)
+
+	// AutoSave is enabled by default.
+	// Now we disable it.
+	e.EnableAutoSave(false)
+
+	var err error
+	logErr := func(action string) {
+		if err != nil {
+			t.Fatalf("test action[%s] failed, err: %v", action, err)
+		}
+	}
+
+	// Because AutoSave is disabled, the policy change only affects the policy in Casbin enforcer,
+	// it doesn't affect the policy in the storage.
+	_, err = e.AddPolicy("alice", "data1", "write")
+	logErr("AddPolicy")
+	// Reload the policy from the storage to see the effect.
+	err = e.LoadPolicy()
+	logErr("LoadPolicy")
+	// This is still the original policy.
+	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}})
+
+	// Now we enable the AutoSave.
+	e.EnableAutoSave(true)
+
+	// Because AutoSave is enabled, the policy change not only affects the policy in Casbin enforcer,
+	// but also affects the policy in the storage.
+	_, err = e.AddPolicy("alice", "data1", "write")
+	logErr("AddPolicy2")
+	// Reload the policy from the storage to see the effect.
+	err = e.LoadPolicy()
+	logErr("LoadPolicy2")
+	// The policy has a new rule: {"alice", "data1", "write"}.
+	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}, {"alice", "data1", "write"}})
+
+	// Remove the added rule.
+	_, err = e.RemovePolicy("alice", "data1", "write")
+	logErr("RemovePolicy")
+	err = e.LoadPolicy()
+	logErr("LoadPolicy3")
+	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}})
+
+	// Remove "data2_admin" related policy rules via a filter.
+	// Two rules: {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"} are deleted.
+	_, err = e.RemoveFilteredPolicy(0, "data2_admin")
+	logErr("RemoveFilteredPolicy")
+	err = e.LoadPolicy()
+	logErr("LoadPolicy4")
+
+	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}})
+}
+
+func testFilteredPolicy(t *testing.T, a *Adapter) {
+	// Initialize some policy in DB.
+	initPolicy(t, a)
+	// Note: you don't need to look at the above code
+	// if you already have a working DB with policy inside.
+
+	// Now the DB has policy, so we can provide a normal use case.
+	// Create an adapter and an enforcer.
+	// NewEnforcer() will load the policy automatically.
+	e, _ := casbin.NewEnforcer("examples/rbac_model.conf")
+	// Now set the adapter
+	e.SetAdapter(a)
+
+	var err error
+	logErr := func(action string) {
+		if err != nil {
+			t.Fatalf("test action[%s] failed, err: %v", action, err)
+		}
+	}
+
+	// Load only alice's policies
+	err = e.LoadFilteredPolicy(Filter{V0: []string{"alice"}})
+	logErr("LoadFilteredPolicy")
+	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}})
+
+	// Load only bob's policies
+	err = e.LoadFilteredPolicy(Filter{V0: []string{"bob"}})
+	logErr("LoadFilteredPolicy2")
+	testGetPolicy(t, e, [][]string{{"bob", "data2", "write"}})
+
+	// Load policies for data2_admin
+	err = e.LoadFilteredPolicy(Filter{V0: []string{"data2_admin"}})
+	logErr("LoadFilteredPolicy3")
+	testGetPolicy(t, e, [][]string{{"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}})
+
+	// Load policies for alice and bob
+	err = e.LoadFilteredPolicy(Filter{V0: []string{"alice", "bob"}})
+	logErr("LoadFilteredPolicy4")
+	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}})
+}
+
+func testRemovePolicies(t *testing.T, a *Adapter) {
+	// Initialize some policy in DB.
+	initPolicy(t, a)
+	// Note: you don't need to look at the above code
+	// if you already have a working DB with policy inside.
+
+	// Now the DB has policy, so we can provide a normal use case.
+	// Create an adapter and an enforcer.
+	// NewEnforcer() will load the policy automatically.
+	e, _ := casbin.NewEnforcer("examples/rbac_model.conf")
+
+	// Now set the adapter
+	e.SetAdapter(a)
+
+	var err error
+	logErr := func(action string) {
+		if err != nil {
+			t.Fatalf("test action[%s] failed, err: %v", action, err)
+		}
+	}
+
+	err = a.AddPolicies("p", "p", [][]string{{"max", "data2", "read"}, {"max", "data1", "write"}, {"max", "data1", "delete"}})
+	logErr("AddPolicies")
+
+	// Load policies for max
+	err = e.LoadFilteredPolicy(Filter{V0: []string{"max"}})
+	logErr("LoadFilteredPolicy")
+
+	testGetPolicy(t, e, [][]string{{"max", "data2", "read"}, {"max", "data1", "write"}, {"max", "data1", "delete"}})
+
+	// Remove policies
+	err = a.RemovePolicies("p", "p", [][]string{{"max", "data2", "read"}, {"max", "data1", "write"}})
+	logErr("RemovePolicies")
+
+	// Reload policies for max
+	err = e.LoadFilteredPolicy(Filter{V0: []string{"max"}})
+	logErr("LoadFilteredPolicy2")
+
+	testGetPolicy(t, e, [][]string{{"max", "data1", "delete"}})
+}
+
+func testAddPolicies(t *testing.T, a *Adapter) {
+	// Initialize some policy in DB.
+	initPolicy(t, a)
+	// Note: you don't need to look at the above code
+	// if you already have a working DB with policy inside.
+
+	// Now the DB has policy, so we can provide a normal use case.
+	// Create an adapter and an enforcer.
+	// NewEnforcer() will load the policy automatically.
+	e, _ := casbin.NewEnforcer("examples/rbac_model.conf")
+
+	// Now set the adapter
+	e.SetAdapter(a)
+
+	var err error
+	logErr := func(action string) {
+		if err != nil {
+			t.Fatalf("test action[%s] failed, err: %v", action, err)
+		}
+	}
+
+	err = a.AddPolicies("p", "p", [][]string{{"max", "data2", "read"}, {"max", "data1", "write"}})
+	logErr("AddPolicies")
+
+	// Load policies for max
+	err = e.LoadFilteredPolicy(Filter{V0: []string{"max"}})
+	logErr("LoadFilteredPolicy")
+
+	testGetPolicy(t, e, [][]string{{"max", "data2", "read"}, {"max", "data1", "write"}})
+}
+
+func testUpdatePolicies(t *testing.T, a *Adapter) {
+	// Initialize some policy in DB.
+	initPolicy(t, a)
+	// Note: you don't need to look at the above code
+	// if you already have a working DB with policy inside.
+
+	// Now the DB has policy, so we can provide a normal use case.
+	// Create an adapter and an enforcer.
+	// NewEnforcer() will load the policy automatically.
+	e, _ := casbin.NewEnforcer("examples/rbac_model.conf")
+
+	// Now set the adapter
+	e.SetAdapter(a)
+
+	var err error
+	logErr := func(action string) {
+		if err != nil {
+			t.Fatalf("test action[%s] failed, err: %v", action, err)
+		}
+	}
+
+	err = a.UpdatePolicy("p", "p", []string{"bob", "data2", "write"}, []string{"alice", "data2", "write"})
+	logErr("UpdatePolicy")
+
+	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"alice", "data2", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}})
+
+	err = a.UpdatePolicies("p", "p", [][]string{{"alice", "data1", "read"}, {"alice", "data2", "write"}}, [][]string{{"bob", "data1", "read"}, {"bob", "data2", "write"}})
+	logErr("UpdatePolicies")
+
+	testGetPolicy(t, e, [][]string{{"bob", "data1", "read"}, {"bob", "data2", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}})
+}
+
+func testUpdateFilteredPolicies(t *testing.T, a *Adapter) {
+	// Initialize some policy in DB.
+	initPolicy(t, a)
+	// Note: you don't need to look at the above code
+	// if you already have a working DB with policy inside.
+
+	// Now the DB has policy, so we can provide a normal use case.
+	// Create an adapter and an enforcer.
+	// NewEnforcer() will load the policy automatically.
+	e, _ := casbin.NewEnforcer("examples/rbac_model.conf")
+
+	// Now set the adapter
+	e.SetAdapter(a)
+
+	e.UpdateFilteredPolicies([][]string{{"alice", "data1", "write"}}, 0, "alice", "data1", "read")
+	e.UpdateFilteredPolicies([][]string{{"bob", "data2", "read"}}, 0, "bob", "data2", "write")
+	e.LoadPolicy()
+	testGetPolicyWithoutOrder(t, e, [][]string{{"alice", "data1", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}, {"bob", "data2", "read"}})
+}
+
+func testGetPolicyWithoutOrder(t *testing.T, e *casbin.Enforcer, res [][]string) {
+	myRes := e.GetPolicy()
+	log.Print("Policy: ", myRes)
+
+	if !arrayEqualsWithoutOrder(myRes, res) {
+		t.Error("Policy: ", myRes, ", supposed to be ", res)
+	}
+}
+
+func arrayEqualsWithoutOrder(a [][]string, b [][]string) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	mapA := make(map[int]string)
+	mapB := make(map[int]string)
+	order := make(map[int]struct{})
+	l := len(a)
+
+	for i := 0; i < l; i++ {
+		mapA[i] = util.ArrayToString(a[i])
+		mapB[i] = util.ArrayToString(b[i])
+	}
+
+	for i := 0; i < l; i++ {
+		for j := 0; j < l; j++ {
+			if _, ok := order[j]; ok {
+				if j == l-1 {
+					return false
+				} else {
+					continue
+				}
+			}
+			if mapA[i] == mapB[j] {
+				order[j] = struct{}{}
+				break
+			} else if j == l-1 {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func TestAdapters(t *testing.T) {
+	a := NewAdapter("tcp", "127.0.0.1:6379")
 	// Use the following if Redis has password like "123"
 	//a := NewAdapterWithPassword("tcp", "127.0.0.1:6379", "123")
 
-	e, _ = casbin.NewEnforcer("examples/rbac_model.conf", a)
-	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}})
-
-	// Add one policy to DB
-	a.AddPolicy("p", "p", []string{"paul", "data2", "read"})
-	e.ClearPolicy()
-	a.LoadPolicy(e.GetModel())
-	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}, {"paul", "data2", "read"}})
-
-	// Remove one policy from DB
-	a.RemovePolicy("p", "p", []string{"paul", "data2", "read"})
-	e.ClearPolicy()
-	a.LoadPolicy(e.GetModel())
-	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}})
-
-	// Add policies to DB
-	a.AddPolicies("p", "p", [][]string{
-		{"curry", "data1", "write"},
-		{"kobe", "data2", "read"},
-	})
-	e.ClearPolicy()
-	a.LoadPolicy(e.GetModel())
-	testGetPolicy(t, e, [][]string{
-		{"alice", "data1", "read"},
-		{"bob", "data2", "write"},
-		{"data2_admin", "data2", "read"},
-		{"data2_admin", "data2", "write"},
-		{"curry", "data1", "write"},
-		{"kobe", "data2", "read"},
-	})
-
-	// Remove polices from DB
-	a.RemovePolicies("p", "p", [][]string{
-		{"curry", "data1", "write"},
-		{"kobe", "data2", "read"},
-	})
-	e.ClearPolicy()
-	a.LoadPolicy(e.GetModel())
-	testGetPolicy(t, e, [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}, {"data2_admin", "data2", "read"}, {"data2_admin", "data2", "write"}})
+	testSaveLoad(t, a)
+	testAutoSave(t, a)
+	testFilteredPolicy(t, a)
+	testAddPolicies(t, a)
+	testRemovePolicies(t, a)
+	testUpdatePolicies(t, a)
+	testUpdateFilteredPolicies(t, a)
 }

